cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ServiceHttpServer.java

```diff
@@ -181,6 +206,22 @@ protected void shutDown() throws Exception {
     }
   }
 
+  public void setInstanceCount(int instanceCount) {
+    this.instanceCount = instanceCount;
+    this.contextFactory = createHttpServiceContextFactory();
```

Comment:
Made both instanceCount and contextFactory volatile. Also moved it to after the final fields.